architecture/2025-03-upgrade-design-doc.md

```diff
@@ -0,0 +1,793 @@
+
+#### Proposed Option
+
+I recommend implementing the helm-based upgrade approach with automatic user data backup/restore functionality (we can even discuss if we would like this functionality in version 1 or not) because:
+
+1. It provides the best balance between user experience simplicity and technical safety
+2. Leverages existing Helm functionality while adding Radius-specific safety features
+3. The automated user data backup/restore mechanism offers critical protection against data loss
+4. Clear progress indicators and health checks give users confidence in the process
+5. This approach aligns with how users currently install/reinstall Radius, providing consistency
+
+### API design (if applicable)
+
+No specific REST API addition is necessary.
+
+### CLI Design (if applicable)
+
+Main CLI changes:
+
+- Addition of `rad upgrade kubernetes` command (details mentioned above)
+- Lock logic will be added to the existing commands that change user data like:
+  - `rad deploy app.bicep`
+  - `rad app delete my-app`
+  - `rad env delete default`
+
+### Implementation Details
+
+The implementation will primarily focus on the following components:
+
+1. **Upgrade Command**: The `rad upgrade kubernetes` command implementation in the CLI codebase
+2. **Version Validation**: Logic to verify compatibility between versions
+3. **Lock Mechanism**: Data-store-level distributed locking system
+4. **Backup/Restore**: User data protection system using ConfigMaps/PVs
+5. **Helm Integration**: Enhanced wrapper around Helm's upgrade capabilities
+6. **Health Verification**: Component readiness and health check mechanisms
+
+All components will follow Radius coding standards and include comprehensive unit tests.
+
+### Error Handling
+
+The upgrade process will implement the following error handling strategies:
+
+1. **Pre-flight Validation**: Catch incompatibility issues before starting the upgrade
+2. **Graceful Timeouts**: All operations will respect user-defined or default timeouts
+3. **Automatic Rollback**: Failed upgrades trigger automatic restoration of previous state
+4. **Detailed Error Reporting**: Clear error messages with troubleshooting guidance
+5. **Idempotent Operations**: Commands can be safely retried after addressing issues
+6. **Resource Cleanup**: Temporary resources created during the upgrade are properly removed
+
+## Test Plan
+
+### Unit Tests
+
+- Test each interface implementation independently
+
+### Integration Tests
+
+- Verify interactions between components during upgrade process
+- Test lock acquisition and release across multiple commands
+- Validate backup/restore operations against actual databases
+
+### End-to-End Tests
+
+- Perform upgrades across consecutive versions (e.g., v0.43 → v0.44)
+- Test version skipping scenarios (e.g., v0.40 → v0.44)
+- Verify custom configuration is properly applied
+- Simulate failures and verify automatic rollback
+
+## Security
+
+No changes to the existing security model needed.
```

Comment:
More details about where the backup is stored, who can access it, when it gets deleted are things that would be relevant here